src/ape/contracts/base.py

```diff
@@ -900,6 +938,46 @@ def _events_(self) -> Dict[str, List[ContractEvent]]:
             # NOTE: Must raise AttributeError for __attr__ method or will seg fault
             raise AttributeError(str(err)) from err
 
+    @cached_property
+    def _errors_(self) -> Dict[str, List[Type[CustomError]]]:
+        abis: Dict[str, List[ErrorABI]] = {}
+
+        try:
+            for abi in self.contract_type.errors:
+                if abi.name in abis:
+                    abis[abi.name].append(abi)
+                else:
+                    abis[abi.name] = [abi]
+
+            # Check for prior error sub-class definitions for the same contract.
+            prior_errors = self.chain_manager.contracts._get_errors(self.address)
+
+            errors = {}
+            for abi_name, abi_list in abis.items():
+                errors_to_add = []
+                for abi in abi_list:
+                    error_type = None
+                    for existing_cls in prior_errors:
+                        if existing_cls.abi.signature == abi.signature:
+                            # Error class was previously initialized by contract at same address.
+                            error_type = existing_cls
```

Comment:
really really important to dynamically-define classes only once if we want to rely on the instances for assertions.... I was seeing weird things in tests when multiple instances of `ContractInstance` with same address / chain would create new classes of custom errors and cause odd test failures like `MyError is not MyError`. So this whole caching system ensures they're only defined one time.